File: typescript/src/giftcard/application/integration/payment/handlers/confirmTopUpFromPayment.ts
import type { PaymentResultIntegrationEvent } from '@/payment/integration/paymentResultIntegrationEvent.js';
import { PaymentResult } from '@/giftcard/application/integration/payment/adapter/paymentResult.js';
import { TopUpConfirming } from '@/giftcard/application/usecases/topUpConfirming.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class ConfirmTopUpFromPayment {
  private readonly _paymentResult: PaymentResult;
  private readonly _useCase: TopUpConfirming;

  constructor(paymentResult: PaymentResult, topUpConfirming: TopUpConfirming) {
    requireDependency(paymentResult, 'paymentResult');
    this._paymentResult = paymentResult;
    requireDependency(topUpConfirming, 'topUpConfirming');
    this._useCase = topUpConfirming;
  }

  handle(event: PaymentResultIntegrationEvent): void {
    const command = this._paymentResult.adapt(event);

    if (command !== null) {
      this._useCase.invoke(command);
    }
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/integration/payment/handlers/ConfirmTopUpFromPayment.java
package it.giannibombelli.wsc2026.giftcard.application.integration.payment.handlers;

import it.giannibombelli.wsc2026.giftcard.application.commands.ConfirmTopUp;
import it.giannibombelli.wsc2026.giftcard.application.integration.payment.adapter.PaymentResult;
import it.giannibombelli.wsc2026.giftcard.application.usecases.TopUpConfirming;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;

import static java.util.Objects.requireNonNull;

public final class ConfirmTopUpFromPayment {
    private final PaymentResult paymentResult;
    private final TopUpConfirming useCase;

    public ConfirmTopUpFromPayment(PaymentResult paymentResult, TopUpConfirming topUpConfirming) {
        this.paymentResult = requireNonNull(paymentResult);
        this.useCase = requireNonNull(topUpConfirming);
    }

    public void handle(PaymentResultIntegrationEvent event) {
        final ConfirmTopUp cmd = paymentResult.adapt(event);

        if (cmd != null) {
            useCase.invoke(cmd);
        }
    }
}
